### PR TITLE
Removing SetZero()

### PR DESCRIPTION
**Pull request by freewym**
resloved issue #1340.

Will run a tdnn recipe for test

### CODE REVIEW

File: src/nnet3/nnet-chain-training.cc [hunk 1]
```diff
@@ -38,9 +38,11 @@ NnetChainTrainer::NnetChainTrainer(const NnetChainTrainingOptions &opts,
   KALDI_ASSERT(opts.nnet_config.momentum >= 0.0 &&
                opts.nnet_config.max_param_change >= 0.0);
   delta_nnet_ = nnet_->Copy();
+  ScaleNnet(0.0, delta_nnet_);
   bool is_gradient = false;  // setting this to true would disable the
                              // natural-gradient updates.
-  SetZero(is_gradient, delta_nnet_);
+  if (is_gradient)
```
Comment (danpovey) [inline src/nnet3/nnet-chain-training.cc]: we just set is_gradient = false, so there is no need to call SetAsGradeintNnet().
Sometimes I declare a variable is_gradient = false just before SetZero() to document what the 'false' means.

File: src/nnet3/nnet-component-itf.h [hunk 2]
```diff
@@ -408,6 +403,10 @@ class UpdatableComponent: public Component {
   /// a different value than x will returned.
   BaseFloat LearningRate() const { return learning_rate_; }
 
+  /// \brief Sets is_gradient_ to true and sets learning_rate_ to 1, ignoring
+  /// learning_rate_factor_.
+  void SetAsGradient() { SetActualLearningRate(1.0); is_gradient_ = true; }
```
Comment (danpovey) [inline src/nnet3/nnet-component-itf.h]: i think doing `learning_rate_ = 1.0;` might be clearer than SetActualLearningRate().

File: src/nnet3/nnet-optimize-test.cc [hunk 3]
```diff
@@ -97,7 +97,9 @@ static bool UnitTestNnetOptimizeWithOptions(int32 srand_seed,
   bool is_gradient = true;  // with natural gradient, the consolidation would
   // affect the final model params -> test just the
   // gradient.
-  SetZero(is_gradient, &nnet_to_update);
+  ScaleNnet(0.0, &nnet_to_update);
+  if (is_gradient)
```
Comment (danpovey) [inline src/nnet3/nnet-optimize-test.cc]: again, you can simplify this code.

File: src/nnet3/nnet-optimize-test.cc [hunk 4]
```diff
@@ -107,7 +109,9 @@ static bool UnitTestNnetOptimizeWithOptions(int32 srand_seed,
   Nnet nnet_opt(nnet);  // copy of the nnet for the optimized computation.
   // necessary in case backprop changes parameters.
   Nnet nnet_opt_to_update(nnet_opt);
-  SetZero(is_gradient, &nnet_opt_to_update);
+  ScaleNnet(0.0, &nnet_opt_to_update);
+  if (is_gradient)
+    SetAsGradientNnet(&nnet_opt_to_update);
```
Comment (danpovey) [inline src/nnet3/nnet-optimize-test.cc]: .. and here too, is_gradient has not changed.

File: src/nnet3/nnet-simple-component.cc [hunk 5]
```diff
@@ -1136,8 +1136,13 @@ void RectifiedLinearComponent::StoreStats(
 }
 
 void AffineComponent::Scale(BaseFloat scale) {
-  linear_params_.Scale(scale);
-  bias_params_.Scale(scale);
+  if (scale == 0.0) {
```
Comment (danpovey) [inline src/nnet3/nnet-simple-component.cc]: you could comment:
// If scale == 0.0 we call SetZero() which will get rid of NaN's and inf's.

File: src/nnet3/nnet-simple-component.cc [hunk 6]
```diff
@@ -4780,8 +4751,19 @@ std::string CompositeComponent::Info() const {
 
 // virtual
 void CompositeComponent::Scale(BaseFloat scale) {
-  for (size_t i = 0; i < components_.size(); i++)
-    components_[i]->Scale(scale);
+  if (scale == 0.0) {
```
Comment (danpovey) [inline src/nnet3/nnet-simple-component.cc]: I don't think you have to modify the code of this function, just leave it as it is.

File: src/nnet3/nnet-utils.cc [hunk 7]
```diff
@@ -280,10 +261,31 @@ void SetLearningRate(BaseFloat learning_rate,
   }
 }
 
+void SetAsGradientNnet(Nnet *nnet) {
```
Comment (danpovey) [inline src/nnet3/nnet-utils.cc]: could you please rename this function to `SetNnetAsGradient()`?

File: src/nnet3/nnet-utils.cc [hunk 8]
```diff
@@ -280,10 +261,31 @@ void SetLearningRate(BaseFloat learning_rate,
...
+      u_comp->SetAsGradient();
+    }
+  }
+}
+
 void ScaleNnet(BaseFloat scale, Nnet *nnet) {
   if (scale == 1.0) return;
   else if (scale == 0.0) {
-    SetZero(false, nnet);
+    for (int32 c = 0; c < nnet->NumComponents(); c++) {
```
Comment (danpovey) [inline src/nnet3/nnet-utils.cc]: I don't think we need to have a special case here any more, for scale == 0.0.
Just make sure that all the components that store stats [which do all support the
Scale() function], have a Scale() function that really sets everything to zero if
called with 0.0.

File: src/nnet3/nnet-component-itf.h (hunk 2 above)
Comment (freewym) [inline src/nnet3/nnet-component-itf.h]: SetActualLearningRate() is a virtual function that was overwritten in CompositeComponent, so we have to call it here.

**Comment by danpovey** [comment]
I think CompositeComponent would have to override SetAsGradient() anyway,
to call the same function on all the owned components.

**Pull request merged**